# Исследование надежности заемщиков


Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.


Цель исследования - ответить на следующие вопросы:
 - Есть ли зависимость между количеством детей и возвратом кредита в срок?
 - Есть ли зависимость между семейным положением и возвратом кредита в срок?
 - Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
 - Как разные цели кредита влияют на его возврат в срок?
 
 Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')<div class="alert alert-info">

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для анализа зависимости сделаем следующие действия:
1. Cоздадим доп. столбец children_count , в котором сделаем категоризацию детей:
   - *нет детей*, если детей нет;
   - *есть дети*, если 1 или 2 ребенка;
   - *многодетный*, если 3 и более детей;
2. Cоздадим сводную таблицу.

In [22]:
# Создадим функцию категоризации по детям
def child_category(child):
    if 1 <= child <= 2:
        return 'есть дети'
    if child >= 3:
        return 'многодетная'
    return 'нет детей'

In [23]:
# Создадим доп. столбец и применим к нему функцию
data['children_count'] = data['children'].apply(child_category)

In [24]:
# Проверим результат по последним 5 значениям датафрейма
data.tail()

children  days_employed  dob_years education  education_id  \
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category children_count  
21520  операции с недвижимостью      есть дети  
21521    операции с автомобилем      нет детей  
21522  операции с недвижимостью      есть дети  
21523    операции с автомобилем    многодетная  
21524    операции с автомобилем      есть дети

*Создадим сводную таблицу методом groupby().*

In [25]:
child_debt = data.groupby('children_count')['debt'].agg(['count','sum'])
child_debt = child_debt.rename(columns={'count':'total','sum':'debt'})

In [26]:
# Считаем долю должников
child_debt['share_of_debt'] = child_debt['debt']/child_debt['total']

In [27]:
# Переводим значения в столбце share_of_debt в проценты и выводим таблицу
child_debt['share_of_debt'] = child_debt['share_of_debt'].map('{:.1%}'.format)
child_debt

total  debt share_of_debt
children_count                           
есть дети        6860   638          9.3%
многодетная       380    31          8.2%
нет детей       14091  1063          7.5%

*Создаим функцию function создающую сводную таблицу методом pivot_table() с выводом на экран, где переменной будет интересующий нас столбец*

In [37]:
def function(varibale):
    df = data.pivot_table(index=varibale, values='debt', aggfunc={'count', 'sum', 'mean'})
    df.columns = ['all_credits', 'share_debtors', 'all_debtors']
    display(df.sort_values(by='share_debtors', ascending=False))

In [38]:
# Вызовем функцию по столбцу children_count:
function(['children_count'])

all_credits  share_debtors  all_debtors
children_count                                         
есть дети            6860.0       0.093003        638.0
многодетная           380.0       0.081579         31.0
нет детей           14091.0       0.075438       1063.0

**Вывод:** Как видно из сводной таблицы наименьшая доля должников у заемщиков без детей (7.5%).
Наибольшая доля должников у заемщиков с 1-м и 2-мя детьми (9.3%), скорее всего это связано с возросшими финансовыми нагрузками.
Гипотеза о зависимости между кол-вом детей и возвратом кредита подтверждается. 
У заемщиков с 3-мя и более детьми доля должников ниже (8.2%), чем у заемщиков с 1 и 2 детьми.
Вероятно, это связано с тем, что существуют программы поддержки многодетных семей. Или количества заемщиков данной категории
недостаточно для данного сравнения.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

*Создадим сводную таблицу методом pivot_table().*

In [28]:
family_status_debt = data.pivot_table(index = ['family_status'], values = 'debt', aggfunc = {'count','sum'})
family_status_debt = family_status_debt.rename(columns={'count':'total','sum':'debt'})

In [29]:
# Считаем долю должников
family_status_debt['share_of_debt'] = family_status_debt['debt']/family_status_debt['total']

In [30]:
# Переводим значения в столбце share_of_debt в проценты и выводим таблицу
family_status_debt['share_of_debt'] = family_status_debt['share_of_debt'].map('{:.1%}'.format)
family_status_debt.sort_values(by='share_of_debt', ascending=False)

total  debt share_of_debt
family_status                                   
Не женат / не замужем   2796   273          9.8%
гражданский брак        4134   385          9.3%
женат / замужем        12261   927          7.6%
в разводе               1189    84          7.1%
вдовец / вдова           951    63          6.6%

In [39]:
# Вызовем функцию по столбцу family_status:
function(['family_status'])

all_credits  share_debtors  all_debtors
family_status                                                 
Не женат / не замужем       2796.0       0.097639        273.0
гражданский брак            4134.0       0.093130        385.0
женат / замужем            12261.0       0.075606        927.0
в разводе                   1189.0       0.070648         84.0
вдовец / вдова               951.0       0.066246         63.0

**Вывод:** Да, действительно. зависимость есть. Люди разведенные или ставшие вдовцом/вдовой, то есть люди состоявшие ранее в браке обладают наименьшими показателями доли должников (7.1% и 6.6% соответственно (Могут быть объеденны в единую категорию)).
Заемщики не сотоящие в браке: не женат/не замужем (9.8%) и состоящие в гражданском браке (9.3%) - имею наивысшие значения доли
должников. Возможно данные категории заемщиков обладают меньшей ответственностью.
Заемщики состоящие в браке являются медианным значением в данной сводной таблице (7.6%). 
Возможные причины: наличие детей, финансовые трудности и т.д.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

*Создадим сводную таблицу методом pivot_table().*

In [31]:
total_income_debt = data.pivot_table(index = [ 'total_income_category'], values = 'debt', aggfunc = {'count','sum'})
total_income_debt = total_income_debt.rename(columns={'count':'total','sum':'debt'})

In [32]:
# Считаем долю должников
total_income_debt['share_of_debt'] = total_income_debt['debt']/total_income_debt['total']

In [33]:
# Переводим значения в столбце share_of_debt в проценты и выводим таблицу
total_income_debt['share_of_debt'] = total_income_debt['share_of_debt'].map('{:.1%}'.format)
total_income_debt.sort_values(by='share_of_debt', ascending=False)

total  debt share_of_debt
total_income_category                           
E                         22     2          9.1%
C                      15921  1353          8.5%
A                         25     2          8.0%
B                       5014   354          7.1%
D                        349    21          6.0%

In [40]:
# Вызовем функцию по столбцу total_income_category:
function(['total_income_category'])

all_credits  share_debtors  all_debtors
total_income_category                                         
E                             22.0       0.090909          2.0
C                          15921.0       0.084982       1353.0
A                             25.0       0.080000          2.0
B                           5014.0       0.070602        354.0
D                            349.0       0.060172         21.0

**Вывод:** Зависимость между уровнем дохода и возвратом кредита в срок есть. 
Количества заемщиков категорий А  и Е (1 000 000 и выше, 0 - 30 000) недостаточно для данного сравнения.
Наибольший показатель доли должников у категории С (50 000 - 200 000) и состовляет 8.5%. Вероятно данная категория
имеет большую финансовую нагрузку при меньшем доходе, чем категории B (200 000 - 1 000 000) с долей задолжности (7.1%)
Самый низкий показатель доли должников у категории D (30 000 - 50 000) - 6%. Скорее всего, из-за низкого дохода, определяющим фактором являются дополнительные расходы, связанные с появлением просрочки оплаты.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

*Создадим сводную таблицу методом groupby().*

In [34]:
purpose_category_debt = data.groupby('purpose_category')['debt'].agg(['count','sum'])
purpose_category_debt = purpose_category_debt.rename(columns={'count':'total','sum':'debt'})

In [35]:
# Считаем долю должников
purpose_category_debt['share_of_debt'] = purpose_category_debt['debt']/purpose_category_debt['total']

In [36]:
# Переводим значения в столбце share_of_debt в проценты и выводим таблицу
purpose_category_debt['share_of_debt'] = purpose_category_debt['share_of_debt'].map('{:.1%}'.format)
purpose_category_debt.sort_values(by='share_of_debt', ascending=False)

total  debt share_of_debt
purpose_category                                   
операции с автомобилем     4279   400          9.3%
получение образования      3988   369          9.3%
проведение свадьбы         2313   183          7.9%
операции с недвижимостью  10751   780          7.3%

In [41]:
# Вызовем функцию по столбцу purpose_category:
function(['purpose_category'])

all_credits  share_debtors  all_debtors
purpose_category                                                 
операции с автомобилем         4279.0       0.093480        400.0
получение образования          3988.0       0.092528        369.0
проведение свадьбы             2313.0       0.079118        183.0
операции с недвижимостью      10751.0       0.072551        780.0

**Вывод:** Наиболее благонадежной целью взятие кредита являются операции с недвижимостью (доля должников наименьшая - 7.3%)
Самые большие доли должников (9.3%) связанны с кредитами на операции с автомобилями и получением образования.


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:*причины пропусков данных в двух столбцах days_employed и total_income (количество пропусков одинаково):
 - неверный ввод данных;
 - сокрытие информации;
 - повреждение файла данных при сохранении;
 - технические проблемы. 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* заполние пропусков медианным значением — лучшее решение для количественных переменных из-за устойчивости медианы к выбросам и аномальным значениям.

### Шаг 4: общий вывод.

В предоставленном наборе данных (размером 21525 строк на 12 колонок) была проведена предобработка данных для иследования по четырем гипотезам, которые и были подтверждены:
1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
4. Как разные цели кредита влияют на его возврат в срок?
- При исследовании данных были обнаружены и удалены строки-дубликаты данных. 
- Так же были найден и удалены аномальные значения в стобце children. 
- Обнаружены пропуски данных в двух столбцах days_employed и total_income (количество пропусков одинаково). Для дальнейшего иследования эти пропуски были заполнены медианными значениями их столбцов.Причинами появления пропусков в данных могут быть: неверный ввод данных, сокрытие информации, повреждение файла данных при сохранении, технические проблемы.
- В столбце education выявлены неявные дубликаты(присутствовали есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв) - приведены к нижнему регистру. 
- В столбце total_income вещественный тип данных заменен на целочисленный.
- Была произведена категоризация данных по столбцам total_income по пяти категориям, purpose по четырем категория, children по трем категориям. Это позволило не только провести более качественный анализ данных, но и выявить недостаток данных для сравнения в некоторых категориях.

По результатам проведенного исследования можно сформировать образ наиболее благонадежного заемщика банка:
 - ранее состоял в браке (в разводе, вдовец / вдова);
 - не имеет детей;
 - со средним доходом от 200 000;
 - берущий кредит на операции с недвижимостью.

Наименее благонадежным заемщиком банка является заемщик:
  - не состоит в браке (не женат / не замужем, гражданский брак);
  - имеет детей (1-го или 2-х);
  - со средним доходом от 50 000 до 200 000;
  - берущий кредит на операции с автомобилем или получение образования.